### ISSUE TITLE
How can it display all available WiFi networks / SSIDs?

### ISSUE THREAD

**Issue by nobody93** [bmegli/wifi-scan#23]
Hi Bartosz, 

Thank you for the very good program. It only displays one SSID, I am wandering if it can display all available WiFi SSIDs like using command line iwlist?

BTW, I believe it won't have memory leak, but I am always a little bit nervous when I see using C malloc and free. Wrap them in C++ class might a good go, using smart pointer will be even better. The hardware MCU is getting cheaper, smaller and powerful, it is not unusual C++ is used in many small IoT devices, but that is my opinion. If you don't mind and if you could also help me, I'll be happy to port it to C++ when I have some time available next month.

Thank you.

Kind regards,

Jupiter

**Comment by bmegli** [comment]
Hi @nobody93,

> Thank you for the very good program. It only displays one SSID, I am wandering if it can display all available WiFi SSIDs like using command line iwlist?

Sure, that's what `wifi-scan-all` example does. See README.md
There is also simplified example in README.md, just right under `wifi-scan-station` example.

Side note:

`wifi-scan-all` triggers a slow passive scan of all networks around. It may not be possible to use it rapidly. Depending on hardware this may mean order of 1 second (when hardware is busy scanning). Also data link may not be usable during this process.

`wifi-scan-station` works in a different way, it may be called rapidly (order of miliseconds)

> BTW, I believe it won't have memory leak, but I am always a little bit nervous when I see using C malloc and free. Wrap them in C++ class might a good go, using smart pointer will be even better. The hardware MCU is getting cheaper, smaller and powerful, it is not unusual C++ is used in many small IoT devices, but that is my opinion. If you don't mind and if you could also help me, I'll be happy to port it to C++ when I have some time available next month.

This is C library on purpose. Feel free to wrap it in C++ for yourself (and other people) if that is your need.

Side note: 

If you are going to use it as library from anything other than C++ you will end up wrapping your C++ interface in C interface again. 

Kind regards,

**Comment by nobody93** [comment]
Thanks Bartosz, based on you experience and opinion, is possible to extend your program to other features:

(1) Switch between WiFi to other network interface to connect to the Internet, if WiFi is available, using WiFi connection, if WiFi is not available, using other network interfaces, very much like a mobile phone, LTE is the default connection, but when WiFi is available, switch to the WiFi.

(2) WiFi C/C++ API for WiFi client to setup and connect to WiFi route by SSID and password? 

Sorry if that question should ask the libmnl.

Thank you.

**Comment by bmegli** [comment]
(1) and (2) - even if it was possible I would not bother. Direct Netlink from C/C++ is rather difficult.

What you ask for is full blown network manager. Look at tools like:
- [Connman](https://en.wikipedia.org/wiki/ConnMan) (this should probably cover 1 and 2)
- whatever is used in Android for network managing
- Ubuntu 18 Netplan
- see how embedded OSes do it (e.g. one example is Connman)
- see how desktop OSes/distros do it (Ubuntu, Arch, Debian, Fedora, CentOS)

TL/DR - there are tools that are meant for this kind of problems. Start your research there.
Only if they don't satisfy your needs (whatever they are - feature set, efficiency, no dependencies, ...) 
think of writing your own.

**Comment by bmegli** [comment]
The original problem was solved, so I am closing the issue.
